buffling/privacy.py

```diff
@@ -0,0 +1,265 @@
+
+  Args: 
+    a, b, c, z: Numeric parameters to the function. 
+    depth: Positive integer, Approximating depth.
+
+  Returns:
+    The approximated value.
+  """
+  k = np.zeros(depth)
+  for i in range(depth): 
+    if i%2==0:
+      variable = i // 2
+      k[i] = (a - c - variable) * (b + variable) 
+    else: 
+      variable = i // 2 + 1
+      k[i] = (b - c - variable) * (a + variable)
+    if c == 0 and i == 0:
+      k[0] /= 1
+    else: 
+      k[i] /= (c + i) * (c + i + 1)
+  f = np.ones(depth)
+  for i in range(depth - 1):
+    f[depth - i - 2] += k[depth - i - 2] * z / f[depth - i - 1]
+  return 1 / f[0]
+
+
+def ratio_of_hypergeometric_mgf(y_d, y_x, p, n_bit, depth=1000):
+  """Compute the ratio of hypergeometric moment-generating functions
+
+  Definition: E((q/p)^(2x)) / E((q/p)^(2y)),
+  where x ~ HG(m, y_d, y_x), y ~ HG(m, y_d + 1, y_x) 
+  Note that the increment is on the second parameter (i.e., y_d + 1).
+
+  Args: 
+    y_d: An integer, the observed # of 1s in bloom filter.
+    y_x: An integer, the # of 1s in bloom filter after "shuffling+blipping".
+    p: An real number in [0,1], bit flipping probability.
+    n_bit: Integer, the length of each bloom filter.
+    depth: Positive integer, Approximating depth. 
+      Will be Passed to evaluate_gauss_continued_fraction().
+
+  Returns:
+    A real number of the simulated epsilon.
+  """
+  q = 1 - p
+  if y_d + y_x > n_bit: 
+    y_d = n_bit - y_d
+    y_x = n_bit - y_x
+  
+  a = - y_d
+  b = - y_x
+  c = n_bit - y_x - y_d 
+  z = (q / p) ** 2
+  l = evaluate_gauss_continued_fraction(a, b, c, z, depth=depth)
+
+  return (n_bit - y_x) / np.maximum(n_bit - y_x - y_d, 1) * l
+
+
+def evaluate_eps_given_yd(y_d, p, n_bit, type="exact", y_x=None, depth=1000):
+  """Evaluate the privacy parameter (epsilon) of one BF.
+
+  Args: 
+    y_d: An integer, the observed # of 1s in bloom filter.
+    p: An real number in [0,1], bit flipping probability.
+    n_bit: Integer, the length of each bloom filter.
+    type: A character string, either "exact" or "asymp".
+    y_x: An integer, the # of 1s in bloom filter after "shuffling+blipping".
+    depth: Integer, Approximating depth. 
+      Will be Passed to evaluate_gauss_continued_fraction().
+
+  Returns:
+    A real number of the simulated epsilon.
+  """
+  ## process the inputs
+  q = 1 - p
+  if y_x == None:
+    y_x = np.random.binomial(y_d, q) + np.random.binomial(n_bit - y_d, p)
+    if y_d + y_x > n_bit: 
+      y_d = n_bit - y_d
+      y_x = n_bit - y_x
+
+  if (type == "asymp"): ## normal approximation
+    b = 2 * np.log(q / p)
+    l = -b * y_d / n_bit + b ** 2 / 2 * y_d * (n_bit - y_d) * (2 * y_x - n_bit) / n_bit ** 3
+    l = np.exp(l)
+  elif (type == "exact"): ## exact form
+    l = ratio_of_hypergeometric_mgf(y_d=y_d, y_x=y_x, p=p, n_bit=n_bit, depth=depth) 
+  
+  ## Pr(A(D')=X) / Pr(A(D)=X)
+  privacy_loss = p / q / l ## version 2
+
+  return np.log(privacy_loss)
+
+
+def find_yx_given_yd_Is_1(p, n_bit, d):
```

Comment:
yx and yd are not clear. Is it possible to make it more explicit?

Reply:
Thanks, I will change the function names.